core/src/main/java/com/dtolabs/rundeck/core/utils/WaitUtils.java

```diff
@@ -0,0 +1,38 @@
+package com.dtolabs.rundeck.core.utils;
+
+import java.time.Duration;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class WaitUtils {
+
+    /**
+     * Waits for the resource to be in the expected state by retrieving it and evaluating its acceptance as many times as needed.
+     *
+     * @param retryableResourceRetriever executed multiple times to retrieve the current state of the R resource.
+     * @param resourceAcceptanceEvaluator evaluates the state of the resource and returns true if the resource is accepted.
+     * @param timeout max duration to wait for the resource to reach the expected state.
+     * @param checkPeriod time to wait between each check.
+     * @return the resource that met the acceptance criteria.
+     * @throws ResourceAcceptanceTimeoutException if the timeout is reached waiting for the resource to reach the expected state.
+     * @throws InterruptedException if the thread sleep was interrupted.
+     */
+    public static <R> R waitFor(
+            Supplier<R> retryableResourceRetriever,
+            Function<R, Boolean> resourceAcceptanceEvaluator,
+            Duration timeout,
+            Duration checkPeriod) throws InterruptedException {
+        R r = retryableResourceRetriever.get();
+        Boolean acceptanceResult = resourceAcceptanceEvaluator.apply(r);
+        long initTime = System.currentTimeMillis();
+        while (!acceptanceResult) {
+            Thread.sleep(Math.min(checkPeriod.toMillis(), timeout.toMillis()));
+            if ((System.currentTimeMillis() - initTime) >= timeout.toMillis()) {
+                throw new ResourceAcceptanceTimeoutException("Timeout reached (" + timeout.toMillis() + "ms) waiting for value: " + r + " to reach the desired state");
+            }
```

Comment:
it seems the logic here will result in: resource acceptance check->false, wait, throw.  So there seems to be an unnecessary wait before the throw.  i.e. it would be better to have a loop like:
- initial check
- wait
    - check again,
    - if true, pass
    - if false,  
        - check timeout expired, if true throw
        - otherwise loop again

Reply:
Indeed, this implementation does not guarantee to timeout at `timeout`. In fact, it cannot do it even in theory if the resource retrieval takes over `timeout`, but let's ignore that possibility for now.

The current implementation attempts to timeout between the `timeout` and `timeout + checkPeriod`.

I'm not sure I fully understand your proposed algo, but I think it still suffers from the possibility of running over the `timeout`. Also, I think it would be logically equivalent to adding another timeout check before the sleep?:
```
            if ((System.currentTimeMillis() - initTime) >= timeout.toMillis()) {
                throw new ResourceAcceptanceTimeoutException("Timeout reached (" + timeout.toMillis() + "ms) waiting for value: " + r + " to reach the desired state");
            }
            Thread.sleep(Math.min(checkPeriod.toMillis(), timeout.toMillis()));
            if ((System.currentTimeMillis() - initTime) >= timeout.toMillis()) {
                throw new ResourceAcceptanceTimeoutException("Timeout reached (" + timeout.toMillis() + "ms) waiting for value: " + r + " to reach the desired state");
            }
```

Does it make sense?

Reply:
Would a "compromise improvement" be to just move the timeout check before the sleep so that it looks something like:

```
        while (!acceptanceResult) {
            if ((System.currentTimeMillis() - initTime) >= timeout.toMillis()) {
                throw new ResourceAcceptanceTimeoutException("Timeout reached (" + timeout.toMillis() + "ms) waiting for value: " + r + " to reach the desired state");
            }
            
            Thread.sleep(Math.min(checkPeriod.toMillis(), timeout.toMillis()));

            r = retryableResourceRetriever.get();
            acceptanceResult = resourceAcceptanceEvaluator.apply(r);
        }
```

Reply:
This can work as well, but there is an increased probability of doing a retreival after the timeout.
Surrounding the `sleep` with checks appears to be a compromise.

Reply:
The code you suggest @ChuckCrawford  is what I was attempting to describe. I was just trying to avoid the unnecessary wait. Yes it could check before and after the wait if you want to be tidy, but really it's not very important for the solution if we overrun the timeout. Either of those options would avoid having an unneeded wait 👍